rfc-091-sharing-assets.md

```diff
@@ -0,0 +1,47 @@
+# Sharing assets
+
+## Summary
+
+We need to share assets to make sure users only have to download a thing once. To do this, we'll upload each application's assets to the same location.
+
+## Problem
+
+In [RFC84 we proposed replacing Static with a gem](https://github.com/alphagov/govuk-rfcs/pull/84). One of the architectural challenges with this approach is that by moving assets (like CSS, JS and images) into the gem, the assets won't be shared for users. This means that users visiting GOV.UK would have to download assets multiple times, depending on what app renders the page.
+
+## Proposal
+
+We upload the assets to a central location. For example, if both `collections` and `government-frontend` have a file named `foo.css`, they will upload the file to `assets.publishing.service.gov.uk/shared/foo.css`.
```

Comment:
The bit that makes me a bit uncomfortable with this is having it be very hard to determine the source of an asset based on the filename.

Eg we'd likely have:
`https://assets.publishing.service.gov.uk/government-frontend/application-00d6906ee26428034a3cc61e0498475b38903f895aeaf5f53222efba375a20c7.css` => `https://assets.publishing.service.gov.uk/shared/application-00d6906ee26428034a3cc61e0498475b38903f895aeaf5f53222efba375a20c7.css`
and
`https://assets.publishing.service.gov.uk/frontend/application-acf275543318182565f911d30e18a81240b22ba94fac1a5b19f9f3c3697b922f.css`
=>
`https://assets.publishing.service.gov.uk/shared/application-acf275543318182565f911d30e18a81240b22ba94fac1a5b19f9f3c3697b922f.css`

but both of those application.css files are very different.

I can't really think of a good centralised way that can be resolved, but it might be nice to have guidelines on the format assets that are publicly exposed are stored so that their paths have more answers.

E.g. we could suggest that assets directories all contain the name of the app (which wouldn't be a concern of a shared resource via a gem) which would then provide clues from the filename of the source file.

Reply:
For application level styles I'd expect them to always be unique and therefore wouldn't need to be hoisted up to a shared layer?

This RFC only needs to apply to shared styles between applications.

Reply:
I'm not aware that this RFC discriminates between a shared or application specific asset - but probably better for @tijmenb to clarify that.

This line seems to imply all assets though: https://github.com/alphagov/govuk-rfcs/pull/91/files#diff-c53d3b658ba990b1192236d1f1fccfe8R23

Reply:
Ah okay, then if all assets are hoisted I agree application bundles might need an exception for clarity.

Reply:
In my head I've assumed that we are talking about keeping some separation between the unique to the app assets and the shared ones that previously came from static.  In order to avoid making users download the same assets multiple times depending on the rendering app for some content we should try to make sure that the assets that come from static aren't asset-pipeline/webpack/whatever smooshed together into one file with the assets that come directly from the rendering app.  So that the browser will ask for `/app-name/application.css` (this could be `/shared/application.css and we rely on the fingerprint to distinguish the versions for each application, but I like the idea of having the app name in there somehow to help with identification - and clearing out old ones) and `/shared/shared.css`.  These `/app-name/` assets will likely always be unique to the app, but the `/shared/`ones shouldn't be (depending on the version, but as we upgrade all apps will converge on the same version) so we get a benefit from sharing them.  Of course, I've only mentioned css files here, but I think the same should apply to js, font, images, etc... too.

Reply:
To be clear, I'm indeed proposing it to work in the way @kevindew describes.

The way I envision this working at the moment is that after Rails precompiles the assets, we will upload the directory (`/public/assets`) to an S3 bucket. We won't touch the source directory. Maybe it feels less uncomfortable to think of the bucket as a kind of cache shared between apps? I'm not sure I have many problems with a bucket full of `application-*.css` files.

Reply:
Ah, ok, so we'll get the benefit of shared versions for things like images and fonts and such, but probably not for JS and CSS bundles, because they'll all be smooshed together by asset-pipeline and each app will probably have their own unique bit of CSS or JS that means their `application.[js|css]` has a different fingerprint, even though 90% of it will be the same as all the other apps?

Actually, I don't know how true that is?  Maybe there aren't many apps that have their own bits of JS + CSS beyond what they currently get from static?  Have we checked?

Reply:
Ha yes - this is what Paul alluded to in https://github.com/alphagov/govuk-rfcs/pull/91#discussion_r166927276. For this to work you we'll need to put only app-specific things in `application.css`. Shared things need to be compiled separately (this already happens with the font, jquery and govuk-template assets).